### PR TITLE
add model registry

### PR DESCRIPTION
**Pull request by xvjiarui**
This PR refactored `Registry` class in following  ways:

1. make `build` function an attribute of `Registry` class.

before:
```python
model = build_from_cfg(cfg, MODELS)
```
after:
```python
model = MODELS.build(cfg)
```
`build` function could be configured by passing argument.  

2. Added hierarchy structure for Registry. 

For example, we can define `MODEL` registry in both [MMDetection](https://github.com/open-mmlab/mmdetection) and [MMClassification](https://github.com/open-mmlab/mmclassification)  as followed:
In MMDetection:

```python
from mmcv.utils import Registry
from mmcv.cnn import MODELS as MMCV_MODELS
MODELS = Registry('model', parent=MMCV_MODELS)

@MODELS.register_module()
class NetA(nn.Module):
    def forward(self, x):
        return x
```

In MMClassification:

```python
from mmcv.utils import Registry
from mmcv.cnn import MODELS as MMCV_MODELS
MODELS = Registry('model', parent=MMCV_MODELS)

@MODELS.register_module()
class NetB(nn.Module):
    def forward(self, x):
        return x + 1
```

We could build either `NetA` or `NetB` by:

```python
from mmcv.cnn import MODELS as MMCV_MODELS
net_a = MMCV_MODELS.build(cfg=dict(type='mmdet.NetA'))
net_b = MMCV_MODELS.build(cfg=dict(type='mmcls.NetB'))
```

# Modification for MM repos:
before:
```python
from mmcv.utils import Registry, build_from_cfg
BACKBONES = Registry('backbone')
def build(cfg, registry, default_args=None):
    """Build a module.

    Args:
        cfg (dict, list[dict]): The config of modules, is is either a dict
            or a list of configs.
        registry (:obj:`Registry`): A registry the module belongs to.
        default_args (dict, optional): Default arguments to build the module.
            Defaults to None.

    Returns:
        nn.Module: A built nn module.
    """
    if isinstance(cfg, list):
        modules = [
            build_from_cfg(cfg_, registry, default_args) for cfg_ in cfg
        ]
        return nn.Sequential(*modules)
    else:
        return build_from_cfg(cfg, registry, default_args)

def build_backbone(cfg):
    """Build backbone."""
    return build(cfg, BACKBONES) 
```
after:
```python
from mmcv.cnn import ModelRegistry
from mmcv.cnn import MODELS as MMCV_MODELS
MODELS = ModelRegistry('models', parent=MMCV_MODELS)
def build_model(cfg):
    """Build model."""
    return MODELS.build(cfg)
```

### CODE REVIEW

**Comment by hellock** [comment]
We can briefly describe the design in the PR description to involve more community discussions.

**Comment by ZwwWayne** [comment]
The original goal of design seems to be achieved perfectly. One remaining question: is there a guide to bypass the 3-level hierarchy? Because from the code [here](https://github.com/open-mmlab/mmcv/pull/760/files#diff-4dd793cf824f1b0688049ed7ce700d16a0709e9f4eb1212c46e73371b5c70db0R106), it seems that only a 2-level hierarchy is accepted.

**Comment by ZwwWayne** [comment]
We may also provide a guide for migrating from the old code to the new code (or what do we need to take care of to avoid BC-breaking) in the doc if necessary.

**Comment by xvjiarui** [comment]
> The original goal of design seems to be achieved perfectly. One remaining question: is there a guide to bypass the 3-level hierarchy? Because from the code [here](https://github.com/open-mmlab/mmcv/pull/760/files#diff-4dd793cf824f1b0688049ed7ce700d16a0709e9f4eb1212c46e73371b5c70db0R106), it seems that only a 2-level hierarchy is accepted.

Multi-level is supported.

**Comment by xvjiarui** [comment]
> We may also provide a guide for migrating from the old code to the new code (or what do we need to take care of to avoid BC-breaking) in the doc if necessary.

I will provide a illustration in this PR. And I will create a PR in model for demo.

File: mmcv/utils/registry.py [hunk 1]
```diff
@@ -28,14 +96,43 @@ def __repr__(self):
...
 
+    @staticmethod
+    def infer_scope():
+        filename = inspect.getmodule(inspect.stack()[2][0]).__name__
+        split_filename = filename.split('.')
+        return split_filename[0]
+
+    @staticmethod
+    def split_scope_key(key):
+        # split_list = key.split('.')
```
Comment (ZwwWayne) [inline mmcv/utils/registry.py:107]: clean commented code.

**Comment by ZwwWayne** [comment]
We may also need to provide one more registry for DATASET, thus downstream tasks can use similar build function from MMCV to build model, dataset, and runner in the unified entry point/engine.

File: mmcv/utils/registry.py [hunk 2]
```diff
@@ -5,16 +5,84 @@
...
 
-    def __init__(self, name):
+    def __init__(self, name, build_func=None, parent=None, scope=None):
         self._name = name
         self._module_dict = dict()
+        self._children = dict()
+        self._scope = self.infer_scope() if scope is None else scope
+        if parent is not None:
+            assert isinstance(parent, Registry)
+            parent._add_children(self)
```
Comment (hellock) [inline mmcv/utils/registry.py:78]: We may do this step at last.

File: mmcv/utils/registry.py [hunk 3]
```diff
@@ -5,16 +5,84 @@
...
+        obj_cls = obj_type
+    else:
+        raise TypeError(
+            f'type must be a str or valid type, but got {type(obj_type)}')
+
+    return obj_cls(**args)
+
+
 class Registry:
     """A registry to map strings to classes.
```
Comment (hellock) [inline mmcv/utils/registry.py:55]: Documentation can be here.

File: mmcv/utils/registry.py [hunk 4]
```diff
@@ -28,14 +96,36 @@ def __repr__(self):
                      f'items={self._module_dict})'
         return format_str
 
+    @staticmethod
+    def infer_scope():
```
Comment (hellock) [inline mmcv/utils/registry.py:100]: Docstring

File: mmcv/utils/registry.py [hunk 5]
```diff
@@ -45,7 +135,27 @@ def get(self, key):
         Returns:
             class: The corresponding class.
         """
-        return self._module_dict.get(key, None)
+        scope, real_key = self.split_scope_key(key)
+        if scope is not None:
+            return self._children[scope].get(real_key)
```
Comment (hellock) [inline mmcv/utils/registry.py:140]: Does this satisfy the desired use case?

File: mmcv/utils/registry.py [hunk 6]
```diff
@@ -45,7 +135,27 @@ def get(self, key):
...
+                # get from children
+                for registry in self._children.values():
+                    result = registry.get(real_key)
+                    if result is not None:
+                        return result
+
+    def build(self, *args, **kwargs):
+        return self.build_func(*args, **kwargs, registry=self)
+
+    def _add_children(self, registry):
```
Comment (hellock) [inline mmcv/utils/registry.py:155]: Docstring.

File: mmcv/utils/registry.py (hunk 5 above)
Comment (xvjiarui) [reply mmcv/utils/registry.py:172]: Yep. The object could be built with inheritance in a recursive way.

File: docs/registry.md [hunk 7]
```diff
@@ -66,5 +67,43 @@ If the module is successfully registered, you can use this converter through con
...
+class NetB(nn.Module):
+    def forward(self, x):
+        return x + 1
+```
+
+We could build either `NetA` or `NetB` by:
+
+```python
+from mmcv.cnn import MODELS as MMCV_MODELS
+net_a = MMCV_MODELS.build(cfg=dict(type='mmdet.NetA'))
```
Comment (ZwwWayne) [inline docs/registry.md:107]: There are two questions that might need to be answered  in the doc. 

1. In this case, can we still build net_a and net_b through

```python
net_a = MMCV_MODELS.build(cfg=dict(type='NetA'))
net_b = MMCV_MODELS.build(cfg=dict(type='NetB'))
```
2. If MMCls and MMDet both create a class named `NetA`, what will happen if we try

```python
net_a = MMCV_MODELS.build(cfg=dict(type='NetA'))
```
Does this depend on the order during registration or something else? Or the conflict can only be resolved by use `mmdet.NetA` and `mmcls.NetA` ?

**Comment by ZwwWayne** [comment]
> We may also need to provide one more registry for DATASET, thus downstream tasks can use similar build function from MMCV to build model, dataset, and runner in the unified entry point/engine.

Ping this part. Are we going to add this in this PR or in the future when necessary?

**Comment by xvjiarui** [comment]
> > We may also need to provide one more registry for DATASET, thus downstream tasks can use similar build function from MMCV to build model, dataset, and runner in the unified entry point/engine.
> 
> Ping this part. Are we going to add this in this PR or in the future when necessary?

The dataset build function is not unified yet. I suggest add DATASET in a new PR>

**Comment by ZwwWayne** [comment]
> > > We may also need to provide one more registry for DATASET, thus downstream tasks can use similar build function from MMCV to build model, dataset, and runner in the unified entry point/engine.
> > 
> > 
> > Ping this part. Are we going to add this in this PR or in the future when necessary?
> 
> The dataset build function is not unified yet. I suggest add DATASET in a new PR>

OK.

File: docs/registry.md (hunk 7 above)
Comment (xvjiarui) [reply docs/registry.md:107]: Currently, it is depend on the order.

**Comment by ZwwWayne** [comment]
LGTM now. See if @hellock has any comments.

**Review by ZwwWayne — APPROVED** [review]

**Comment by xvjiarui** [comment]
Kindly ping @hellock.

**Comment by ZwwWayne** [comment]
May merge conflicts first.

File: mmcv/utils/registry.py [hunk 8]
```diff
@@ -28,14 +107,57 @@ def __repr__(self):
...
+            >>> Registry.split_scope_key('mmdet.ResNet')
+            'mmdet', 'ResNet'
+            >>> Registry.split_scope_key('ResNet')
+            None, 'ResNet'
+
+        Return:
+            scope (str, None): The first scope.
+            key (str): The remaining key.
+        """
+        split_index = key.find('.')
```
Comment (hellock) [inline mmcv/utils/registry.py:139]: What if there are more than one dot in `key`?

Comment (xvjiarui) [reply mmcv/utils/registry.py:139]: Handled in a recursive way.

File: mmcv/utils/registry.py [hunk 9]
```diff
@@ -45,7 +179,42 @@ def get(self, key):
         Returns:
             class: The corresponding class.
         """
-        return self._module_dict.get(key, None)
+        scope, real_key = self.split_scope_key(key)
+        if scope is not None:
+            if scope in self._children:
```
Comment (hellock) [inline mmcv/utils/registry.py:184]: What if users define two registries in one codebases and make one registry be the parent of the other?

```python
A = Registry('models')
B = Registry('sub_models', parent=A)
```

Comment (xvjiarui) [reply mmcv/utils/registry.py:194]: The scope of B could be inferred automatically.

File: mmcv/utils/registry.py [hunk 10]
```diff
@@ -5,16 +5,98 @@
...
+            Default: None.
     """
 
-    def __init__(self, name):
+    def __init__(self, name, build_func=None, parent=None, scope=None):
         self._name = name
         self._module_dict = dict()
+        self._children = dict()
+        self._scope = self.infer_scope() if scope is None else scope
+        if build_func is None:
```
[the remark this answers is not in the document]
Comment (xvjiarui) [reply mmcv/utils/registry.py:95]: Added. It also explained in the docstring.

**Comment by xvjiarui** [comment]
There is a remaining issue, how could build modules in `mmcv.MODELS` with `mmdet.MODELS`? What prefix should be used? 
Currently, if `mmdet.MODELS.build` get `"ResNet"`, it will search `mmdet.MODELS` only. If `mmdet.MODELS.build` get `"mmaction.ResNet3D"` it will search `mmcv.MODELS` from top to bottom.

File: docs/registry.md [hunk 11]
```diff
@@ -71,5 +72,63 @@ If the module is successfully registered, you can use this converter through con
 
 ```python
 converter_cfg = dict(type='Converter1', a=a_value, b=b_value)
-converter = build_converter(converter_cfg)
+converter = CONVERTERS.build(converter_cfg)
+```
+
+## Hierarchy Registry
```
Comment (hellock) [inline docs/registry.md:78]: This part needs refactoring. Users will get confused by this doc and do not know why and how.

File: docs/registry.md [hunk 12]
```diff
@@ -71,5 +72,63 @@ If the module is successfully registered, you can use this converter through con
...
+@MODELS.register_module()
+class NetB(nn.Module):
+    def forward(self, x):
+        return x + 1
+```
+
+We could build two net in either MMDetection or MMClassification by:
+
+```python
+import mmcls # import mmcls to register models
```
Comment (hellock) [inline docs/registry.md:111]: Could this be omitted? It is unexpected that users need to modify the source codes apart from setting the configs.

File: mmcv/cnn/builder.py [hunk 13]
```diff
@@ -0,0 +1,29 @@
+import torch.nn as nn
+
+from ..utils import Registry, build_from_cfg
+
+
+def build_model_from_cfg(cfg, registry, default_args=None):
+    """Build a module from config dict(s).
```
Comment (hellock) [inline mmcv/cnn/builder.py:7]: Explain the difference from `build_from_cfg`.

File: mmcv/utils/registry.py [hunk 14]
```diff
@@ -5,16 +5,106 @@
...
         name (str): Registry name.
+        build_func(func, optional): Build function to construct instance from
+            Registry, func:`build_from_cfg` is used if neither ``parent`` or
+            ``build_func`` is specified. If ``parent`` is specified and
+            ``build_func`` is not given,  ``build_func`` will be inherited
+            from ``parent``. Default: None.
+        parent (Registry, optional): Parent registry. The class registered in
+            children registry could be built from parent. Default: None.
+        scope (str, optional): The scope of registry. It is the key to index
+            children. If not specified, scope will be the name of the package
```
Comment (hellock) [inline mmcv/utils/registry.py:81]: What does it mean by "It is the key to index children"?

**Comment by xvjiarui** [comment]
[ ] Refactor registry.md

**Comment by xvjiarui** [comment]
@hellock doc refactor done.

**Review by ZwwWayne — APPROVED** [review]

**Comment by xvjiarui** [comment]
@hellock another doc refactor done.

**Review by hellock — APPROVED** [review]

**Pull request merged**